File: bolt-qt/src/daemonclient.cpp
#include "daemonclient.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QStandardPaths>
#include <unistd.h>

DaemonClient::DaemonClient(QObject *parent)
    : QObject(parent)
    , m_socket(new QLocalSocket(this))
    , m_pollTimer(new QTimer(this))
    , m_reconnectTimer(new QTimer(this))
{
    m_pollTimer->setInterval(1000);
    m_reconnectTimer->setInterval(3000);
    m_reconnectTimer->setSingleShot(true);

    connect(m_socket, &QLocalSocket::connected, this, &DaemonClient::onSocketConnected);
    connect(m_socket, &QLocalSocket::disconnected, this, &DaemonClient::onSocketDisconnected);
    connect(m_socket, &QLocalSocket::errorOccurred, this, &DaemonClient::onSocketError);
    connect(m_socket, &QLocalSocket::readyRead, this, &DaemonClient::onReadyRead);
    connect(m_pollTimer, &QTimer::timeout, this, &DaemonClient::poll);
    connect(m_reconnectTimer, &QTimer::timeout, this, &DaemonClient::tryConnect);

    // Defer initial connection to the event loop so consumers can
    // connect to our signals before we potentially emit connected().
    QTimer::singleShot(0, this, &DaemonClient::tryConnect);
}

QString DaemonClient::socketPath() {
    QString runtimeDir = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
    if (!runtimeDir.isEmpty())
        return runtimeDir + "/bolt/bolt.sock";
    return QString("/tmp/bolt-%1/bolt.sock").arg(getuid());
}

void DaemonClient::tryConnect() {
    if (m_socket->state() != QLocalSocket::UnconnectedState)
        return;
    QString path = socketPath();
    // Ensure we connect to the filesystem socket, not abstract namespace
    m_socket->setSocketOptions(QLocalSocket::NoOptions);
    m_socket->setServerName(path);
    m_socket->connectToServer(QIODevice::ReadWrite);
}

void DaemonClient::onSocketConnected() {
    m_connected = true;
    m_reconnectTimer->stop();
    m_pollTimer->start();
    resetParserState();
    emit connected();
    fetchDownloads();
}

void DaemonClient::onSocketDisconnected() {
    m_connected = false;
    m_pollTimer->stop();
    m_requestInFlight = false;
    m_pollInFlight = false;
    m_queue.clear();
    resetParserState();
    m_reconnectTimer->start();
    emit disconnected();
}

void DaemonClient::onSocketError(QLocalSocket::LocalSocketError) {
    if (!m_connected && !m_reconnectTimer->isActive())
        m_reconnectTimer->start();
}

void DaemonClient::resetParserState() {
    m_headerBuffer.clear();
    m_bodyBuffer.clear();
    m_contentLength = -1;
    m_responseStatusCode = 0;
    m_headersComplete = false;
}

void DaemonClient::sendRequest(const QByteArray &method, const QByteArray &path,
                                const QByteArray &body, const QString &tag) {
    PendingRequest req;
    req.method = method;
    req.path = path;
    req.body = body;
    req.tag = tag;
    m_queue.enqueue(req);
    processQueue();
}

void DaemonClient::processQueue() {
    if (m_requestInFlight || m_queue.isEmpty() || !m_connected)
        return;

    m_currentRequest = m_queue.dequeue();
    m_requestInFlight = true;
    resetParserState();

    QByteArray request;
    request.append(m_currentRequest.method);
    request.append(" ");
    request.append(m_currentRequest.path);
    request.append(" HTTP/1.1\r\n");
    request.append("Host: localhost\r\n");

    if (!m_currentRequest.body.isEmpty()) {
        request.append("Content-Type: application/json\r\n");
        request.append("Content-Length: ");
        request.append(QByteArray::number(m_currentRequest.body.size()));
        request.append("\r\n");
    }

    request.append("\r\n");
    request.append(m_currentRequest.body);

    m_socket->write(request);
    m_socket->flush();
}

void DaemonClient::onReadyRead() {
    while (m_socket->bytesAvailable() > 0) {
        if (!m_headersComplete) {
            m_headerBuffer.append(m_socket->readAll());
            int headerEnd = m_headerBuffer.indexOf("\r\n\r\n");
            if (headerEnd < 0)
                return;

            // Parse status line
            int firstLineEnd = m_headerBuffer.indexOf("\r\n");
            QByteArray statusLine = m_headerBuffer.left(firstLineEnd);
            // "HTTP/1.1 200 OK"
            int spaceIdx = statusLine.indexOf(' ');
            if (spaceIdx >= 0) {
                int secondSpace = statusLine.indexOf(' ', spaceIdx + 1);
                QByteArray codeStr = (secondSpace >= 0)
                    ? statusLine.mid(spaceIdx + 1, secondSpace - spaceIdx - 1)
                    : statusLine.mid(spaceIdx + 1);
                m_responseStatusCode = codeStr.toInt();
            }

            // Parse Content-Length
            QByteArray headers = m_headerBuffer.left(headerEnd);
            m_contentLength = 0;
            for (const QByteArray &line : headers.split('\n')) {
                QByteArray trimmed = line.trimmed();
                if (trimmed.toLower().startsWith("content-length:")) {
                    m_contentLength = trimmed.mid(15).trimmed().toInt();
                    break;
                }
            }

            // Move remaining data to body buffer
            m_bodyBuffer = m_headerBuffer.mid(headerEnd + 4);
            m_headerBuffer.clear();
            m_headersComplete = true;
        } else {
            m_bodyBuffer.append(m_socket->readAll());
        }

        if (m_headersComplete && m_bodyBuffer.size() >= m_contentLength) {
            QByteArray body = m_bodyBuffer.left(m_contentLength);
            QByteArray remaining = m_bodyBuffer.mid(m_contentLength);

            QString tag = m_currentRequest.tag;
            int statusCode = m_responseStatusCode;

            m_requestInFlight = false;
            resetParserState();

            // If there's leftover data, keep it for next response
            if (!remaining.isEmpty())
                m_headerBuffer = remaining;

            handleResponse(statusCode, body, tag);
            processQueue();
        }
    }
}

void DaemonClient::handleResponse(int statusCode, const QByteArray &body, const QString &tag) {
    QJsonDocument doc = QJsonDocument::fromJson(body);
    QJsonObject obj = doc.object();

    // Handle error responses
    if (statusCode >= 400) {
        QString errorMsg = obj["error"].toString();
        QString errorCode = obj["code"].toString();

        // Special case: probe failure
        if (tag == "probe") {
            emit probeFailed(errorMsg);
            return;
        }

        // Special case: add download duplicate
        if (tag == "addDownload" && errorCode == "DUPLICATE_FILENAME") {
            emit requestFailed(tag, statusCode, errorCode, errorMsg);
            return;
        }

        emit requestFailed(tag, statusCode, errorCode, errorMsg);

        // Clear poll flag if this was a poll request
        if (tag == "poll")
            m_pollInFlight = false;
        return;
    }

    // Route by tag
    if (tag == "poll" || tag == "fetchDownloads") {
        QJsonArray arr = obj["downloads"].toArray();
        QVector<Download> downloads;
        downloads.reserve(arr.size());
        for (const QJsonValue &v : arr)
            downloads.append(Download::fromJson(v.toObject()));
        if (tag == "poll")
            m_pollInFlight = false;
        emit downloadsFetched(downloads);
    } else if (tag == "addDownload") {
        Download dl = Download::fromJson(obj["download"].toObject());
        emit downloadAdded(dl);
        fetchDownloads();
    } else if (tag == "probe") {
        ProbeResult result = ProbeResult::fromJson(obj);
        emit probeCompleted(result);
    } else if (tag == "fetchConfig") {
        Config cfg = Config::fromJson(obj);
        emit configFetched(cfg);
    } else if (tag == "updateConfig") {
        emit configUpdated();
    } else if (tag == "fetchStats") {
        Stats stats = Stats::fromJson(obj);
        emit statsFetched(stats);
    } else if (tag == "pause" || tag == "resume" || tag == "retry" || tag == "delete") {
        // Action succeeded — refresh download list
        if (!m_pollInFlight)
            fetchDownloads();
    }
}

void DaemonClient::poll() {
    if (m_pollInFlight)
        return;
    m_pollInFlight = true;
    sendRequest("GET", "/api/downloads", {}, "poll");
}

// --- Public API methods ---

void DaemonClient::fetchDownloads() {
    sendRequest("GET", "/api/downloads", {}, "fetchDownloads");
}

void DaemonClient::addDownload(const AddRequest &req) {
    QJsonDocument doc(req.toJson());
    sendRequest("POST", "/api/downloads", doc.toJson(QJsonDocument::Compact), "addDownload");
}

void DaemonClient::deleteDownload(const QString &id, bool deleteFile) {
    QByteArray path = "/api/downloads/" + id.toUtf8();
    if (deleteFile)
        path.append("?delete_file=true");
    sendRequest("DELETE", path, {}, "delete");
}

void DaemonClient::pauseDownload(const QString &id) {
    sendRequest("POST", "/api/downloads/" + id.toUtf8() + "/pause", {}, "pause");
}

void DaemonClient::resumeDownload(const QString &id) {
    sendRequest("POST", "/api/downloads/" + id.toUtf8() + "/resume", {}, "resume");
}

void DaemonClient::retryDownload(const QString &id) {
    sendRequest("POST", "/api/downloads/" + id.toUtf8() + "/retry", {}, "retry");
}

void DaemonClient::probeUrl(const QString &url) {
    QJsonObject obj;
    obj["url"] = url;
    QJsonDocument doc(obj);
    sendRequest("POST", "/api/probe", doc.toJson(QJsonDocument::Compact), "probe");
}

void DaemonClient::fetchConfig() {
    sendRequest("GET", "/api/config", {}, "fetchConfig");
}

void DaemonClient::updateConfig(const QJsonObject &partial) {
    QJsonDocument doc(partial);
    sendRequest("PUT", "/api/config", doc.toJson(QJsonDocument::Compact), "updateConfig");
}

void DaemonClient::fetchStats() {
    sendRequest("GET", "/api/stats", {}, "fetchStats");
}

File: bolt-qt/src/bolt_qt/daemon_client.py
from __future__ import annotations

import http.client
import json
import os
import socket
from dataclasses import dataclass
from queue import Empty, Queue

from PySide6.QtCore import QObject, QThread, QTimer, Signal

from .types import Config, Download, ProbeResult, Segment, Stats


def _socket_path() -> str:
    runtime_dir = os.environ.get("XDG_RUNTIME_DIR", "")
    if not runtime_dir:
        runtime_dir = f"/tmp/bolt-{os.getuid()}"
    return os.path.join(runtime_dir, "bolt", "bolt.sock")


class _UnixHTTPConnection(http.client.HTTPConnection):
    def __init__(self, socket_path: str):
        super().__init__("localhost")
        self._socket_path = socket_path

    def connect(self):
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(self._socket_path)
        self.sock = sock


_SILENT_TAGS = frozenset(("poll", "fetchDownloads", "fetchStats", "fetchConfig"))


@dataclass
class _RequestItem:
    method: str
    path: str
    body: bytes
    tag: str


class _WorkerThread(QThread):
    response_ready = Signal(str, int, bytes)  # tag, status_code, body
    connection_changed = Signal(bool)  # True = connected, False = disconnected

    def __init__(self, socket_path: str, parent: QObject | None = None):
        super().__init__(parent)
        self._socket_path = socket_path
        self._queue: Queue[_RequestItem | None] = Queue()
        self._stop = False
        self._conn: _UnixHTTPConnection | None = None

    def enqueue(self, item: _RequestItem) -> None:
        self._queue.put(item)

    def request_stop(self) -> None:
        self._stop = True
        self._queue.put(None)  # unblock the get()

    def run(self) -> None:
        while not self._stop:
            try:
                item = self._queue.get(timeout=0.2)
            except Empty:
                continue
            if item is None:
                continue
            self._execute(item)

        if self._conn:
            try:
                self._conn.close()
            except OSError:
                pass

    def _ensure_connection(self) -> _UnixHTTPConnection:
        if self._conn is None:
            conn = _UnixHTTPConnection(self._socket_path)
            conn.connect()
            self._conn = conn
            self.connection_changed.emit(True)
        return self._conn

    def _close_connection(self) -> None:
        if self._conn:
            try:
                self._conn.close()
            except OSError:
                pass
            self._conn = None

    def _execute(self, item: _RequestItem) -> None:
        headers = {"Host": "localhost", "Connection": "keep-alive"}
        if item.body:
            headers["Content-Type"] = "application/json"
            headers["Content-Length"] = str(len(item.body))

        for attempt in range(2):
            try:
                conn = self._ensure_connection()
                conn.request(item.method, item.path, item.body or None, headers)
                resp = conn.getresponse()
                body = resp.read()
                self.response_ready.emit(item.tag, resp.status, body)
                return
            except (OSError, http.client.HTTPException):
                self._close_connection()
                if attempt == 0:
                    continue
                self.connection_changed.emit(False)
                # Fail the request
                self.response_ready.emit(item.tag, 0, b"")


class DaemonClient(QObject):
    connected = Signal()
    disconnected = Signal()
    downloads_fetched = Signal(list)
    download_added = Signal(object)
    probe_completed = Signal(object)
    probe_failed = Signal(str)
    config_fetched = Signal(object)
    config_updated = Signal()
    stats_fetched = Signal(object)
    download_detail_fetched = Signal(object, list)
    request_failed = Signal(str, int, str, str)  # endpoint, status, code, message

    def __init__(self, parent: QObject | None = None):
        super().__init__(parent)
        self._connected = False
        self._poll_in_flight = False
        self._socket_path = _socket_path()

        self._worker = _WorkerThread(self._socket_path)
        self._worker.response_ready.connect(self._handle_response)
        self._worker.connection_changed.connect(self._on_connection_changed)
        self._worker.start()

        self._poll_timer = QTimer(self)
        self._poll_timer.setInterval(1000)
        self._poll_timer.timeout.connect(self._poll)

        self._reconnect_timer = QTimer(self)
        self._reconnect_timer.setInterval(3000)
        self._reconnect_timer.setSingleShot(True)
        self._reconnect_timer.timeout.connect(self._try_connect)

        # Defer initial connection
        QTimer.singleShot(0, self._try_connect)

    def is_connected(self) -> bool:
        return self._connected

    def shutdown(self) -> None:
        self._poll_timer.stop()
        self._reconnect_timer.stop()
        self._worker.request_stop()
        self._worker.wait(3000)

    # --- Connection management ---

    def _try_connect(self) -> None:
        self._worker.enqueue(_RequestItem("GET", "/api/downloads", b"", "poll"))

    def _on_connection_changed(self, is_connected: bool) -> None:
        if is_connected and not self._connected:
            self._connected = True
            self._reconnect_timer.stop()
            self._poll_timer.start()
            self.connected.emit()
        elif not is_connected and self._connected:
            self._connected = False
            self._poll_in_flight = False
            self._poll_timer.stop()
            if not self._reconnect_timer.isActive():
                self._reconnect_timer.start()
            self.disconnected.emit()
        elif not is_connected and not self._connected:
            # Failed to connect
            if not self._reconnect_timer.isActive():
                self._reconnect_timer.start()
            self.disconnected.emit()

    # --- Response handling ---

    def _handle_response(self, tag: str, status_code: int, body: bytes) -> None:
        # Connection failure
        if status_code == 0:
            if tag == "probe":
                self.probe_failed.emit("Connection lost")
            elif tag not in _SILENT_TAGS:
                self.request_failed.emit(tag, 0, "CONNECTION_LOST", "Connection lost")
            if tag == "poll":
                self._poll_in_flight = False
            return

        obj: dict = {}
        if body:
            try:
                obj = json.loads(body)
            except (json.JSONDecodeError, ValueError):
                pass

        # Error responses
        if status_code >= 400:
            error_msg = obj.get("error", "")
            error_code = obj.get("code", "")
            if tag == "probe":
                self.probe_failed.emit(error_msg)
            else:
                self.request_failed.emit(tag, status_code, error_code, error_msg)
            if tag == "poll":
                self._poll_in_flight = False
            return

        # Route by tag
        if tag in ("poll", "fetchDownloads"):
            downloads = [Download.from_json(d) for d in obj.get("downloads") or []]
            if tag == "poll":
                self._poll_in_flight = False
            self.downloads_fetched.emit(downloads)
        elif tag == "addDownload":
            dl = Download.from_json(obj.get("download", {}))
            self.download_added.emit(dl)
            self.fetch_downloads()
        elif tag == "probe":
            result = ProbeResult.from_json(obj)
            self.probe_completed.emit(result)
        elif tag == "fetchConfig":
            cfg = Config.from_json(obj)
            self.config_fetched.emit(cfg)
        elif tag == "updateConfig":
            self.config_updated.emit()
        elif tag == "fetchStats":
            stats = Stats.from_json(obj)
            self.stats_fetched.emit(stats)
        elif tag == "fetchDetail":
            dl = Download.from_json(obj.get("download", {}))
            segments = [Segment.from_json(s) for s in obj.get("segments") or []]
            self.download_detail_fetched.emit(dl, segments)
        elif tag in ("pause", "resume", "retry", "delete", "pauseAll", "resumeAll", "reorder"):
            if not self._poll_in_flight:
                self.fetch_downloads()

    # --- Polling ---

    def _poll(self) -> None:
        if self._poll_in_flight:
            return
        self._poll_in_flight = True
        self._worker.enqueue(_RequestItem("GET", "/api/downloads", b"", "poll"))

    # --- Public API ---

    def fetch_downloads(self) -> None:
        self._worker.enqueue(_RequestItem("GET", "/api/downloads", b"", "fetchDownloads"))

    def add_download(self, req) -> None:
        body = json.dumps(req.to_json()).encode()
        self._worker.enqueue(_RequestItem("POST", "/api/downloads", body, "addDownload"))

    def delete_download(self, dl_id: str, delete_file: bool) -> None:
        path = f"/api/downloads/{dl_id}"
        if delete_file:
            path += "?delete_file=true"
        self._worker.enqueue(_RequestItem("DELETE", path, b"", "delete"))

    def pause_download(self, dl_id: str) -> None:
        self._worker.enqueue(_RequestItem("POST", f"/api/downloads/{dl_id}/pause", b"", "pause"))

    def resume_download(self, dl_id: str) -> None:
        self._worker.enqueue(_RequestItem("POST", f"/api/downloads/{dl_id}/resume", b"", "resume"))

    def retry_download(self, dl_id: str) -> None:
        self._worker.enqueue(_RequestItem("POST", f"/api/downloads/{dl_id}/retry", b"", "retry"))

    def pause_all(self) -> None:
        self._worker.enqueue(_RequestItem("POST", "/api/downloads/pause-all", b"", "pauseAll"))

    def resume_all(self) -> None:
        self._worker.enqueue(_RequestItem("POST", "/api/downloads/resume-all", b"", "resumeAll"))

    def reorder_downloads(self, ordered_ids: list[str]) -> None:
        body = json.dumps({"ordered_ids": ordered_ids}).encode()
        self._worker.enqueue(_RequestItem("PUT", "/api/downloads/reorder", body, "reorder"))

    def probe_url(self, url: str) -> None:
        body = json.dumps({"url": url}).encode()
        self._worker.enqueue(_RequestItem("POST", "/api/probe", body, "probe"))

    def fetch_config(self) -> None:
        self._worker.enqueue(_RequestItem("GET", "/api/config", b"", "fetchConfig"))

    def update_config(self, partial: dict) -> None:
        body = json.dumps(partial).encode()
        self._worker.enqueue(_RequestItem("PUT", "/api/config", body, "updateConfig"))

    def fetch_stats(self) -> None:
        self._worker.enqueue(_RequestItem("GET", "/api/stats", b"", "fetchStats"))

    def fetch_download_detail(self, dl_id: str) -> None:
        self._worker.enqueue(_RequestItem("GET", f"/api/downloads/{dl_id}", b"", "fetchDetail"))
